homeassistant/components/nasweb/nasweb_data.py

```diff
@@ -0,0 +1,68 @@
+"""Dataclass storing integration data in hass.data[DOMAIN]."""
+
+from dataclasses import dataclass, field
+import logging
+
+from aiohttp.hdrs import METH_POST
+
+from homeassistant.components.webhook import (
+    async_generate_id,
+    async_register as webhook_register,
+    async_unregister as webhook_unregister,
+)
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.network import NoURLAvailableError, get_url
+
+from .const import DOMAIN, WEBHOOK_URL
+from .coordinator import NASwebCoordinator, NotificationCoordinator
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@dataclass
+class NASwebData:
+    """Class storing integration data."""
+
+    notify_coordinator: NotificationCoordinator = field(
+        default_factory=NotificationCoordinator
+    )
+    webhook_id = ""
+    entries_coordinators: dict[str, NASwebCoordinator] = field(default_factory=dict)
+
+    def is_initialized(self) -> bool:
+        """Return True if instance was initialized and is ready for use."""
+        return len(self.webhook_id) > 0
+
+    def can_be_deinitialized(self) -> bool:
+        """Return whether this instance can be deinitialized."""
+        return not self.notify_coordinator.has_coordinators()
+
+    def initialize(self, hass: HomeAssistant) -> None:
+        """Initialize NASwebData instance."""
+        if self.is_initialized():
+            return
+        new_webhook_id = async_generate_id()
+        webhook_register(
+            hass,
+            DOMAIN,
+            "NASweb",
+            new_webhook_id,
+            self.notify_coordinator.handle_webhook_request,
+            allowed_methods=[METH_POST],
+        )
+        self.webhook_id = new_webhook_id
+        _LOGGER.debug("Registered webhook: %s", self.webhook_id)
+
+    def deinitialize(self, hass: HomeAssistant) -> None:
+        """Deinitialize NASwebData instance."""
+        if not self.is_initialized():
+            return
+        webhook_unregister(hass, self.webhook_id)
+
+    def get_webhook_url(self, hass: HomeAssistant) -> str | None:
+        """Return webhook url for Push API."""
+        try:
+            hass_url = get_url(hass)
```

Comment:
Does the device have any restrictions on what URLs it will support? Based on the error messages in `strings.json`, we want the internal URL. In that case, we should set `allow_external` to `False` here.